src/compiler/checker.ts

```diff
@@ -41471,6 +41471,11 @@ namespace ts {
                 }
             }
 
+            if (isForInOrOfStatement(forInOrOfStatement) && isIdentifier(forInOrOfStatement.initializer) && forInOrOfStatement.initializer.escapedText === "async") {

```

Comment:
`for(async in ...)` / `for await ( async of ...)` is legal.

Reply:
Maybe it's better to check it in the parser.

Reply:
Given my experience with the parser, and the comments on the similar PR in acornjs, I suspect this is a better place.

Reply:
should be checked in parser because `for ((async) of []);` is a valid statement. 
https://github.com/babel/babel/issues/13207

Reply:
No, it's easier to let the parser accept `async` like any other identifier, and just exclude the single incorrect case in the grammar checks in the checker. The parser would have to decide on a node to return in the error case, and if it *weren't* `Identifier` then the checker and emitter would have to handle that error case.